### PR TITLE
[Merged by Bors] - feat: port Analysis.InnerProductSpace.l2Space

### PR DESCRIPTION
**Pull request by j-loreaux**
---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Review by Parcly-Taxel — APPROVED** [review]

**Comment by j-loreaux** [comment]
@Parcly-Taxel can you explain your changes? You are undoing a bunch of things I did explicitly as part of the porting process.

**Comment by j-loreaux** [comment]
Please no one merge this right now.

**Comment by Parcly-Taxel** [comment]
I had independently tried porting this file, but did not push because I was stuck on the same universe error you were stuck on. When you pushed your PR and eventually solved the universe error, I decided to weave my version of the file with yours, trying to minimise the diff between both your version and the mathport output.

File: Mathlib/Analysis/InnerProductSpace/l2Space.lean [hunk 1]
```diff
@@ -91,9 +91,6 @@ open scoped BigOperators NNReal ENNReal Classical ComplexConjugate Topology
 
 noncomputable section
 
-local macro_rules | `($x ^ $y)   => `(HPow.hPow $x $y)
--- porting note: see lean4#2220
-
```
Comment (Parcly-Taxel) [inline Mathlib/Analysis/InnerProductSpace/l2Space.lean]: I found out that this hack wasn't needed anywhere in the file, and it isn't needed to solve the universe error, so in the interest of minimising the diff with the mathport output I have removed this.

**Comment by semorrison** [comment]
But @Parcly-Taxel, so many of the changes in your last commit are unhelpful. You've removed comments and porting notes that Jireh added. :-(

File: Mathlib/Analysis/InnerProductSpace/Orthogonal.lean [hunk 2]
```diff
@@ -138,7 +138,7 @@ theorem isClosed_orthogonal : IsClosed (Kᗮ : Set E) := by
 #align submodule.is_closed_orthogonal Submodule.isClosed_orthogonal
 
 /-- In a complete space, the orthogonal complement of any submodule `K` is complete. -/
-instance instOrthogonalCompleteSpace [CompleteSpace E] : CompleteSpace Kᗮ :=
+instance Orthogonal.completeSpace [CompleteSpace E] : CompleteSpace Kᗮ :=
```
Comment (semorrison) [inline Mathlib/Analysis/InnerProductSpace/Orthogonal.lean:141]: We're currently (but far from consistently) aiming to have `inst` at the beginning of all manually named instances.

File: Mathlib/Analysis/InnerProductSpace/l2Space.lean [hunk 3]
```diff
@@ -104,8 +101,6 @@ variable {G : ι → Type _} [∀ i, NormedAddCommGroup (G i)] [∀ i, InnerProd
 
 local notation "⟪" x ", " y "⟫" => @inner 𝕜 _ _ x y
 
-/-- `ℓ²(ι, 𝕜)` is the Hilbert space of square-summable functions `ι → 𝕜`, herein implemented
-as `lp (fun i : ι => 𝕜) 2`. -/
```
Comment (semorrison) [inline Mathlib/Analysis/InnerProductSpace/l2Space.lean]: Restore this?

File: Mathlib/Analysis/InnerProductSpace/l2Space.lean [hunk 4]
```diff
@@ -119,25 +114,26 @@ theorem summable_inner (f g : lp G 2) : Summable fun i => ⟪f i, g i⟫ := by
...
-        _ = ∑' i, re ⟪f i, f i⟫ := by
-          congr
-          funext i
-          rw [norm_sq_eq_inner (𝕜 := 𝕜)]
-          -- porting note: `simp` couldn't do this anymore
-        _ = re (∑' i, ⟪f i, f i⟫) := (IsROrC.reClm.map_tsum ?_).symm
-      · norm_num
-      · exact summable_inner f f
+      -- Porting note: all but the last step of the `calc` block moved into a `have`
+      -- so as to not fail
```
Comment (semorrison) [inline Mathlib/Analysis/InnerProductSpace/l2Space.lean:125]: Revert this change? The previous version doesn't use the `have ... convert`.

File: Mathlib/Analysis/InnerProductSpace/l2Space.lean (hunk 3 above)
Comment (Parcly-Taxel) [reply Mathlib/Analysis/InnerProductSpace/l2Space.lean]: Yes, I'm going to do so. The rationale behind removing this, even though it fails the `docBlame` linter, is that `docBlame` is not in the CI lint.

File: Mathlib/Analysis/InnerProductSpace/l2Space.lean [hunk 5]
```diff
@@ -258,15 +254,14 @@ protected theorem linearIsometry_apply_dfinsupp_sum_single (W₀ : Π₀ i : ι,
 `E` into E, has range the closure of the span of the subspaces. -/
 protected theorem range_linearIsometry [∀ i, CompleteSpace (G i)] :
     LinearMap.range hV.linearIsometry.toLinearMap =
-      (⨆ i, LinearMap.range (V i).toLinearMap).topologicalClosure := by
-    -- porting note: dot notation broken
+    (⨆ i, LinearMap.range (V i).toLinearMap).topologicalClosure := by
```
Comment (semorrison) [inline Mathlib/Analysis/InnerProductSpace/l2Space.lean]: @Parcly-Taxel, please restore the porting note.

File: Mathlib/Analysis/InnerProductSpace/l2Space.lean [hunk 6]
```diff
@@ -290,11 +286,8 @@ space `E` and an orthogonal family `V : Π i, G i →ₗᵢ[𝕜] E` such that t
 
 Keeping in mind that `lp G 2` is "the" external Hilbert sum of `G : ι → Type*`, this is analogous
 to `DirectSum.IsInternal`, except that we don't express it in terms of actual submodules. -/
-structure IsHilbertSum : Prop where
-  ofSurjective ::
-  /-- The orthogonal family consituting the summands in the Hilbert sum. -/
```
Comment (semorrison) [inline Mathlib/Analysis/InnerProductSpace/l2Space.lean]: @Parcly-Taxel, please don't delete doc-strings??

File: Mathlib/Analysis/InnerProductSpace/l2Space.lean [hunk 7]
```diff
@@ -290,11 +286,8 @@ space `E` and an orthogonal family `V : Π i, G i →ₗᵢ[𝕜] E` such that t
...
+structure IsHilbertSum : Prop where ofSurjective ::
   protected OrthogonalFamily : OrthogonalFamily 𝕜 G V
-  /-- The isometry `lp G 2 → E` induced by the orthogonal family is surjective. -/
```
Comment (semorrison) [inline Mathlib/Analysis/InnerProductSpace/l2Space.lean]: And here.

File: Mathlib/Analysis/InnerProductSpace/l2Space.lean [hunk 8]
```diff
@@ -412,7 +403,6 @@ variable (ι) (𝕜) (E)
 /-- A Hilbert basis on `ι` for an inner product space `E` is an identification of `E` with the `lp`
 space `ℓ²(ι, 𝕜)`. -/
 structure HilbertBasis where ofRepr ::
-  /-- The linear isometric equivlence implementing identifiying the hilbert space with `ℓ²`. -/
```
Comment (semorrison) [inline Mathlib/Analysis/InnerProductSpace/l2Space.lean]: And here.

File: Mathlib/Analysis/InnerProductSpace/l2Space.lean [hunk 9]
```diff
@@ -433,10 +423,10 @@ protected theorem repr_symm_single (b : HilbertBasis ι 𝕜 E) (i : ι) :
   rfl
 #align hilbert_basis.repr_symm_single HilbertBasis.repr_symm_single
 
--- porting note: removed `@[simp]` because `simp` can prove this
+-- @[simp] -- Porting note: simp can prove this
 protected theorem repr_self (b : HilbertBasis ι 𝕜 E) (i : ι) :
     b.repr (b i) = lp.single 2 i (1 : 𝕜) := by
-  simp
+  rw [← b.repr_symm_single, LinearIsometryEquiv.apply_symm_apply]
```
Comment (semorrison) [inline Mathlib/Analysis/InnerProductSpace/l2Space.lean]: Why change this?

File: Mathlib/Analysis/InnerProductSpace/l2Space.lean [hunk 10]
```diff
@@ -575,13 +559,13 @@ protected def mkOfOrthogonalEqBot (hsp : (span 𝕜 (Set.range v))ᗮ = ⊥) : H
 #align hilbert_basis.mk_of_orthogonal_eq_bot HilbertBasis.mkOfOrthogonalEqBot
 
 @[simp]
-protected theorem coe_mkOfOrthogonalEqBot (hsp : (span 𝕜 (Set.range v))ᗮ = ⊥) :
+protected theorem coe_of_orthogonal_eq_bot_mk (hsp : (span 𝕜 (Set.range v))ᗮ = ⊥) :
```
Comment (semorrison) [inline Mathlib/Analysis/InnerProductSpace/l2Space.lean]: @Parcly-Taxel, please revert, I think Jireh changed this correctly. :-(

File: Mathlib/Analysis/NormedSpace/lpSpace.lean [hunk 11]
```diff
@@ -714,7 +714,7 @@ section NormedSpace
 
 variable {𝕜 : Type _} [NormedField 𝕜] [∀ i, NormedSpace 𝕜 (E i)]
 
-instance instNormedSpace [Fact (1 ≤ p)] : NormedSpace 𝕜 (lp E p) where
+instance normedSpace [Fact (1 ≤ p)] : NormedSpace 𝕜 (lp E p) where
```
Comment (semorrison) [inline Mathlib/Analysis/NormedSpace/lpSpace.lean:717]: @Parcly-Taxel, please revert, see above.

File: Mathlib/Analysis/InnerProductSpace/l2Space.lean (hunk 10 above)
Comment (Parcly-Taxel) [reply Mathlib/Analysis/InnerProductSpace/l2Space.lean]: Done.

**Comment by Parcly-Taxel** [comment]
@semorrison you may now delete your comments on the code.

**Comment by semorrison** [comment]
> @semorrison you may now delete your comments on the code.

I don't think there's anything to delete: if you followed my suggestions, then clicking "Resolve conversation" was the correct thing to do, and we're now ready for @j-loreaux to have a look over this again.

**Comment by semorrison** [comment]
@Parcly-Taxel, force pushing was pretty unhelpful, as it means there's no history for me to review your reversions... :-(

**Comment by semorrison** [comment]
Gah, you just reverted all of your changes?! The whole point of me reviewing was that many of your changes were good, and I just wanted some fixed.

**Comment by Parcly-Taxel** [comment]
> @Parcly-Taxel, force pushing was pretty unhelpful, as it means there's no history for me to review your reversions... :-(

All I did was to delete my own commit (and just that commit) from the history. Now you know that I reverted all of my changes.

**Comment by Parcly-Taxel** [comment]
> Okay. As a general principle, please don't force push when other people are reviewing your contributions. It makes life difficult, and causes additional reviewing headaches. Given that I had already read all your changes, and implicitly approved all the ones that I didn't ask to be reverted, it seems we're now going in circles.

I still read all your comments and changed my version to incorporate all of them. _Now_ it should be fine :)

File: Mathlib/Analysis/InnerProductSpace/l2Space.lean (hunk 3 above)
Comment (j-loreaux) [reply Mathlib/Analysis/InnerProductSpace/l2Space.lean]: That's no reason to remove a helpful docstring? But, I'll be glad it's back.

**Comment by fpvandoorn** [comment]
@Parcly-Taxel for next time, when you want to do undo a commit, please use `git revert` to undo it. This will make a new commit that undoes a previous commit. This way we keep all the history. This is especially important after someone has reviewed anything on the branch, or if someone else might be working on the same branch.

**Comment by fpvandoorn** [comment]
I merged this branch into the commit that @semorrison reviewed (without actually making any changes to this branch). You can view the changes since the review until now here:
https://github.com/leanprover-community/mathlib4/compare/27cee9e19...31d7053a6/

**Comment by j-loreaux** [comment]
@semorrison the PR is fine with me now. Thanks.

@Parcly-Taxel First, thanks for the improvements, the code formatting was definitely welcome. One very minor point, for unused variables, I try to leave the names in, like `_hi` instead of `_`, because otherwise Lean autonames them, and the generated names might be meaningless in the infoview, which can be confusing when reading a proof. Admittedly, this is partially a matter of taste, some may prefer `_`.

For future reference:

1. when you start working on a file, please PR it as WIP immediately. I don't feel like duplicating work you've done, and I presume vice versa? If you see that I'm porting a file, please don't jump in unless I mark it as `help-wanted` (or it's been a sufficient length of time, like at least 24 hours minimum?). I may have local changes that I just haven't pushed because I am still working on the file (this goes under the duplication-of-effort issue also). Moreover, it's often harder to incorporate two individuals' changes than it is to suggest changes later during the review process. If you want to make changes to something I haven't marked as `help-wanted` just DM me on Zulip to ask.
2. Floris' comment above about `git revert` is the correct way to go. If you think I'm being hypocritical because I force-pushed this branch before you, let me explain: I force-pushed *before* I marked it as awaiting review, and explicitly because I knew it would error during bors without the latest master, and rebasing for porting commits (prior to the review process starting) makes reviewing easier because the diff is better than merging.
3. Please, never delete docstrings; if they are poorly written, then just improve them. Deleting documentation is rarely, if ever, helpful.
4. Same goes for porting notes (although there can be exceptions to this if things change, e.g., lean4#2074 and eta)

**Comment by eric-wieser** [comment]
> Please, never delete docstrings; if they are poorly written, then just improve them. Deleting documentation is rarely, if ever, helpful.

I assume the docstrings in question are ones you added when porting @j-loreaux, not ones from mathport.

I think the tension here is whether the goal of porting is "make the file compile with minimal changes", or whether it's fine to do some cleanup when porting. While it would be "simpler" to use the former approach,I think pragmatically docstrings generally fall on the "fine to cleanup" end of this scale, but I don't know if we ever wrote that down on the porting wiki.

**Comment by eric-wieser** [comment]
We had some problems earlier in the port with people adding docstrings to files they didn't understand just because `#lint` told them too, and as a result ending up with terrible docstrings. I don't for a moment believe that applies to this PR!

**Comment by semorrison** [comment]
bors merge

**Comment by j-loreaux** [comment]
@eric-wieser I understand people adding docstrings they don't understand might be unhelpful, but still I think the solution should not be "*remove* the docstrings", but rather "*improve* the docstrings". That being said, I can understand turning off the linter for CI during the port to reduce the occurrences of people adding bad docstrings that have to be corrected during review.

**Pull request closed without merging**